grout/filters.py

```diff
@@ -68,19 +80,48 @@ def filter_polygon_id(self, queryset, field_name, poly_uuid):
         #    'SELECT geom FROM grout_boundarypolygon WHERE uuid=%s', (poly_uuid,)
         #))
 
-    def filter_record_type(self, queryset, field_name, value):
-        """ Method filter for records having a desired record type (uuid)
-
-        e.g. /api/records/?record_type=44a51b83-470f-4e3d-b71b-e3770ec79772
-
-        """
-        return queryset.filter(schema__record_type=value)
+    def filter_occurred_min(self, queryset, field_name, value):
+        """Add a lower bound for datetime ranges."""
+        if not value:
+            # Provide a hardcoded minimum date of 1AD.
+            min_date = parse('0001-01-01T00:00:00+00:00')
+        else:
+            try:
+                min_date = parse(value)
+            except ValueError:
+                # The parser could not parse the date string, so raise an error.
+                raise QueryParameterException('occurred_min', DATETIME_FORMAT_ERROR)
+
+        if not min_date.tzinfo:
+            raise QueryParameterException('occurred_min', DATETIME_FORMAT_ERROR)
+        else:
+            return queryset.filter(occurred_from__gte=min_date)\
```

Comment:
Your point about the behavior of the `occurred_min`/`occurred_max` filters is a really good one. Thinking about it more closely, it seems like the question the filters are currently answering is "Is the date range entirely contained within the filter range?" Perhaps a more intuitive question for the filter to answer would be "Does any part of the date range overlap with the filter range?"

Formally, this would amount to a change from

```
(from >= min AND to >= min) AND (from <= max AND to <= max)
```

to something like

```
from <= max AND to >= min
```

Does this logic seem sound? I think you're right that there will always be applications where one filter is preferable to the other, so this will definitely need to be well-documented. But I think as long as we're defining the default functionality, the "overlap" filter is probably a more sensible default than the "containment" filter, particularly since it's more forgiving of the end user.

Reply:
Hmm, assuming that `to >= from` (side thought: we should probably add validation for that if it doesn't exist already), I think that the first and second version are functionally identical. Perhaps you meant `from <= max OR to >= min`? I think that would match the written description you gave (which I agree with). Regardless, yeah, I think either way is reasonable, but the more flexible one will probably result in the fewest surprises for new users so that seems like a pretty good default.

Reply:
Agreed that we should validate `to >= from`! Seems simple enough to add in this PR.

OK, so this is definitely not critical to the discussion at hand, but in case it's interesting to you: I drew some pictures to help me think through it and I still think we need both `from <= max` and `to >= min` in order for the condition to hold. As an example, if `from < max` while `to < min` (which would satisfy `from <= max OR to >= min` but not `from <= max AND to >= min`), then a range like this one passes even though there's no overlap:

```
from               to
 |-----------------|
                                x          x
                               min        max

```

I don't think the first two conditions are equal either. Here's an example of a valid "overlap" case where `from <= max AND to >= min` passes, but the original "containment" condition (`(from >= min AND to >= min) AND (from <= max AND to <= max)`) doesn't, because `from < min)`:

```
from               to
 |-----------------|
                x          x
               min        max

```

This overlap also doesn't pass the "containment" condition, because `from < min` and `to > max`, but it does pass both conditions we suggested (`from <= min AND/OR to >= max`):

```
from               to
 |-----------------|
     x          x
    min        max
```

This wouldn't be the first time that range comparisons have got my brain tangled up, so it's entirely possible I've missed something important here!

Reply:
You're exactly right! I've run through this exact problem (with pictures) a few times before so I foolishly thought I could get away with not drawing pictures this time, but nope! :+1: